text/0513-app-specific-staging.md

```diff
@@ -0,0 +1,281 @@
+    * Bootstrapping will be faster since the heavy resource of a KMS key is no longer involved.
+    * Because roles are a global resource, every account now only needs to be bootstrapped once. First of all the lack
+      of necessary control of regions will work a lot better with Control Tower+automatic Stack Sets (which does not
+      allow region control).
+
+If we can make the bootstrapping resources part of the CDK application, then users now have a familiar way to customize
+them to their heart’s content, so the treadmill of bootstrap stack customization requests is going to disappear, and
+customers will also not need to customize the bootstrap template anymore (assuming their customizations have to do with
+the resources instead of the roles).
+
+A downside is potentially that we lose the ability to have a version number on the bootstrapped resources (because SSM
+is not global), but we might say that’s not necessary anymore since the Roles are unlikely to change often.
+
+> If we wanted to maintain versioning on the Roles, we could say that the stack always must be deployed in `us-east-1`
+> and that’s where we look for the version; however, this may require cross-internet traffic and therefore be considered
+> dodgy from a reliability perspective, and we could only do the versioning check using the CLI, not from the
+> CloudFormation template.  Of course we’ll have to pick the correct leader region per partition, `aws-cn`, `aws-iso`, etc.
+
+### How it will work in practice
+
+Bootstrapping resources are currently designed the way they are because the CLI relies on the assumption that the
+bootstrap resources are present with a well-known name, before the first CloudFormation deployment starts. In other
+words, this is purely a limitation of the orchestration, that we can take away.
+
+Here’s what we’re going to do:
+
+* We will introduce a new Stack Synthesizer, called `AppStagingSynthesizer`.
+* This synthesizer will create a support stack with the bucket, and an ECR repository per Docker image.
+* Assets will have a dependency on the support stack. This is a new concept that doesn’t currently exist because assets
+  are an orchestration artifact that looks independent like stacks are, but they aren't really: in practice the orchestration
+  ignores everything except stacks, and treats assets as being part of a stack.
+    * Docker assets may still be built before the first deployment (although for proper caching we need the repository
+      to exist first), but will only be uploaded when it’s their time in the orchestration workflow.
+* For a minimal diff these resources could have fixed names, but we could add support for Stack Outputs and assets could
+  have support for Parameters, so that we can thread generated bucket and repository names through the system. For now,
+  we will do fixed names for the staging resources.
```

Comment:
Is this going to cause issues for cleaning up resources? If I deploy a test app and then delete it am I going to have a bunch of left over repos and buckets that I have to clean up before I can deploy again?